### ISSUE TITLE
Clarify typing of default values in `@data`

### ISSUE THREAD

**Issue by akirakyle** [Roger-luo/Moshi.jl#33]
More playing around and something I'm confused about is whether it's required to always use the `.Type` when creating self-referential data types or not, especially when using default constructors. This is illustrated with the following example:

```julia
using Moshi.Data: @data

@data SExpr begin
    struct Add
        arguments::Vector{SExpr} = SExpr[]
    end
end

SExpr.Add()
```
This errors with `ERROR: MethodError: no method matching getindex(::Module)`

Meanwhile replacing with either
```julia
        arguments::Vector{SExpr} = SExpr.Type[]
```
or
```julia
        arguments::Vector{SExpr.Type} = SExpr.Type[]
```
seems to work fine giving a `Main.SExpr.var"typeof(SExpr)"(Main.SExpr.var"##Storage#Add"(Main.SExpr.var"typeof(SExpr)"[]))`.

It seems like for consistency, if `SExpr` is allowed to refer to the type in the definition of self-referential data types, the it shoud also be allowed in the constructor for the default value?

**Comment by Roger-luo** [comment]
yeah, I agree. This should be considered as a bug - all symbol `SExpr` in the `@data` code block should be referring to `SExpr.Type`. I won't have time working on this recently (credit to US visa scam!) unfortunately, but feel free to figure out a fix you can.
